pkg/controller/mizar/mizar-pod-controller.go

```diff
@@ -0,0 +1,208 @@
+		// Two different versions of the same object will always have different RVs.
+		return
+	}
+	if curObj.DeletionTimestamp != nil {
+		// Object is being deleted. Don't handle update anymore.
+		return
+	}
+
+	key, _ := controller.KeyFunc(curObj)
+	c.queue.Add(KeyWithEventType{Key: key, EventType: "update"})
+}
+
+func (c *MizarPodController) deleteObj(obj interface{}) {
+	key, _ := controller.KeyFunc(obj)
+	klog.Infof("%v deleted. key %s.", controllerForMizarPod, key)
+	c.queue.Add(KeyWithEventType{Key: key, EventType: "delete"})
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the handler is never invoked concurrently with the same key.
+func (c *MizarPodController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *MizarPodController) processNextWorkItem() bool {
+	workItem, quit := c.queue.Get()
+
+	if quit {
+		return false
+	}
+
+	keyWithEventType := workItem.(KeyWithEventType)
+	key := keyWithEventType.Key
+	defer c.queue.Done(key)
+
+	err := c.handler(keyWithEventType)
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("Handle %v of key %v failed with %v", controllerForMizarPod, key, err))
+	c.queue.AddRateLimited(keyWithEventType)
+
+	return true
+}
+
+func (c *MizarPodController) handle(keyWithEventType KeyWithEventType) error {
+	key := keyWithEventType.Key
+	eventType := keyWithEventType.EventType
+	klog.Infof("Entering handling for %v. key %s, eventType %s", controllerForMizarPod, key, eventType)
+
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished handling %v %q (%v)", controllerForMizarPod, key, time.Since(startTime))
+	}()
+
+	tenant, namespace, name, err := cache.SplitMetaTenantNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	obj, err := c.lister.PodsWithMultiTenancy(namespace, tenant).Get(name)
+	if err != nil {
+		return err
+	}
+
+	klog.V(4).Infof("Handling %v %s/%s/%s hashkey %v for event %v", controllerForMizarPod, tenant, namespace, obj.Name, obj.HashKey, eventType)
+
+	switch eventType {
+	case "create":
+		processGrpcReturnCode(c, GrpcCreatePod(c.grpcHost, obj), keyWithEventType)
+	case "update":
+		processGrpcReturnCode(c, GrpcUpdatePod(c.grpcHost, obj), keyWithEventType)
+	case "delete":
+		processGrpcReturnCode(c, GrpcDeletePod(c.grpcHost, obj), keyWithEventType)
+	default:
+		panic(fmt.Sprintf("unimplemented for eventType %v", eventType))
+	}
+
+	return nil
+}
+
+func processGrpcReturnCode(c *MizarPodController, returnCode *ReturnCode, keyWithEventType KeyWithEventType) {
+	key := keyWithEventType.Key
+	switch returnCode.Code {
+	case CodeType_OK:
+		klog.Infof("Mizar handled request successfully for %v. key %s, eventType %s", controllerForMizarPod, key, keyWithEventType.EventType)
+	case CodeType_TEMP_ERROR:
+		klog.Infof("Mizar hit temporary error for %v. key %s. %s, eventType %s", controllerForMizarPod, key, returnCode.Message, keyWithEventType.EventType)
```

Comment:
Not sure whether it is OK to log the error as "temporary". When we say "temporary", it is usually an error that will go after retrial.